### ISSUE TITLE
Wrong type casting google docs

### ISSUE THREAD

**Issue by rockandruby** [roo-rb/roo#498]
When I upload google docs table in xlsx, I get  1920.0 instead 1920. Tested on latest gem version. Also if I save this document locallyt(without any modification, just ctrl+s) I get correct value X. Any ideas?

**Comment by rockandruby** [comment]
after some research I can tell that type casting is correct and I get string type but anyway my string value transforms to X.0 instead X. Also if I save this documen locallyt(without any modification, just ctrl+s) I get correct value X. Any ideas?

**Comment by rockandruby** [comment]
seems the issue in nokogiri, because when I invoke `extract_cells` I already have wrong values

**Comment by chopraanmol1** [comment]
@rockandruby nokogiri doesn't typecast the value. Roo tries to typecast in following method:
https://github.com/roo-rb/roo/blob/a9026fdceebd06ee113bc2bb02f6b0542bc159b3/lib/roo/excelx/cell/number.rb#L19-L29

Please share a sample file which reproduce this issue.

**Comment by rockandruby** [comment]
@chopraanmol1 Yes, roo tries to type cast, but it gets data from nokogiri and in that method wrong data is already present(e.g. 1.0) and type casting goes wrong because `number.include?('.') == true`. In `extract_cells` method if you type `cell_xml.content` u'll get float instead int. [Sample file](https://ufile.io/bmgrf)

**Comment by chopraanmol1** [comment]
@rockandruby I'm unable to download the sample file with the given link. Can you upload as a comment itself? 

I don't think it is an issue with nokogiri gem. You should try reading raw XML file It will also probably have float value. A lot of data like integer, float, date, date-time, boolean are natively stored as a number in the raw XML file and respective excel client consider cell format to determine how to display or in our scenario typecast/convert the value.

I think the issue is with roo not recognizing the particular cell format being used in your sample file. So If you can upload a sample file (as a comment) it will be easy to debug this issue further

**Comment by rockandruby** [comment]
@chopraanmol1 weird, because I can download via that link. Ok, below is a table with following format
First Name | First Name | First Name | First Name | First Name |  
-- | -- | -- | -- | -- | --
One | Two | BTN | 1 | 2019 |  
By the way I parsed objects which nokogiri creates when parsing google xlsx file and noticed that objects with wrong values don't have some attributes. Here's a [link](https://github.com/sparklemotion/nokogiri/issues/1878) to my issue

**Comment by rockandruby** [comment]
Yes, I investigated a bit more and found out that a raw xml file with wrong type casting is given to nokogiri and nokogiri parses it. So we need to figure out why roo creates wrong xml file with numbers and type casts them to float for google docs. Below I attached that xml file. Pay attention to D and E columns. In my xlsx file they're integers but in xml for some reason they're parsed to float.
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:mx="http://schemas.microsoft.com/office/mac/excel/2008/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:mv="urn:schemas-microsoft-com:mac:vml" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main" xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac" xmlns:xm="http://schemas.microsoft.com/office/excel/2006/main"><sheetPr><outlinePr summaryBelow="0" summaryRight="0"/></sheetPr><sheetViews><sheetView workbookViewId="0"/></sheetViews><sheetFormatPr customHeight="1" defaultColWidth="14.43" defaultRowHeight="15.75"/><sheetData><row r="1"><c r="A1" s="1" t="s"><v>0</v></c><c r="B1" s="1" t="s"><v>1</v></c><c r="C1" s="1" t="s"><v>2</v></c><c r="D1" s="1" t="s"><v>3</v></c><c r="E1" s="1" t="s"><v>4</v></c><c r="F1" s="1" t="s"><v>5</v></c><c r="G1" s="1" t="s"><v>6</v></c><c r="H1" s="1" t="s"><v>7</v></c><c r="I1" s="1" t="s"><v>8</v></c><c r="J1" s="1" t="s"><v>9</v></c><c r="K1" s="1" t="s"><v>10</v></c><c r="L1" s="1" t="s"><v>11</v></c><c r="M1" s="1" t="s"><v>12</v></c><c r="N1" s="1" t="s"><v>13</v></c><c r="O1" s="1" t="s"><v>14</v></c><c r="P1" s="1" t="s"><v>15</v></c><c r="Q1" s="1" t="s"><v>16</v></c><c r="R1" s="1" t="s"><v>17</v></c><c r="S1" s="1" t="s"><v>18</v></c><c r="T1" s="1" t="s"><v>19</v></c><c r="U1" s="1" t="s"><v>20</v></c></row><row r="2"><c r="A2" s="2" t="s"><v>21</v></c><c r="B2" s="2" t="s"><v>22</v></c><c r="C2" s="3" t="s"><v>23</v></c><c r="D2" s="2"><v>1.0</v></c><c r="E2" s="2"><v>2019.0</v></c><c r="F2" s="4"/><c r="G2" s="4"/><c r="H2" s="4"/><c r="I2" s="4"/><c r="J2" s="4"/><c r="K2" s="4"/><c r="L2" s="4"/><c r="M2" s="4"/><c r="N2" s="4"/><c r="O2" s="3" t="s"><v>24</v></c><c r="P2" s="4"/><c r="Q2" s="4"/><c r="R2" s="4"/><c r="S2" s="4"/><c r="T2" s="4"/><c r="U2" s="4"/></row><row r="3"><c r="A3" s="2" t="s"><v>21</v></c><c r="B3" s="2" t="s"><v>22</v></c><c r="C3" s="3" t="s"><v>25</v></c><c r="D3" s="2"><v>2.0</v></c><c r="E3" s="2"><v>2019.0</v></c><c r="F3" s="4"/><c r="G3" s="4"/><c r="H3" s="4"/><c r="I3" s="4"/><c r="J3" s="4"/><c r="K3" s="4"/><c r="L3" s="4"/><c r="M3" s="4"/><c r="N3" s="4"/><c r="O3" s="3" t="s"><v>26</v></c><c r="P3" s="4"/><c r="Q3" s="4"/><c r="R3" s="4"/><c r="S3" s="4"/><c r="T3" s="4"/><c r="U3" s="4"/></row><row r="4"><c r="A4" s="2" t="s"><v>21</v></c><c r="B4" s="2" t="s"><v>22</v></c><c r="C4" s="3" t="s"><v>27</v></c><c r="D4" s="2"><v>3.0</v></c><c r="E4" s="2"><v>2019.0</v></c><c r="F4" s="4"/><c r="G4" s="4"/><c r="H4" s="4"/><c r="I4" s="4"/><c r="J4" s="4"/><c r="K4" s="4"/><c r="L4" s="4"/><c r="M4" s="4"/><c r="N4" s="4"/><c r="O4" s="3" t="s"><v>28</v></c><c r="P4" s="4"/><c r="Q4" s="4"/><c r="R4" s="4"/><c r="S4" s="4"/><c r="T4" s="4"/><c r="U4" s="4"/></row><row r="5"><c r="A5" s="2" t="s"><v>21</v></c><c r="B5" s="2" t="s"><v>22</v></c><c r="C5" s="5" t="s"><v>29</v></c><c r="D5" s="2"><v>4.0</v></c><c r="E5" s="2"><v>2019.0</v></c><c r="F5" s="4"/><c r="G5" s="4"/><c r="H5" s="4"/><c r="I5" s="4"/><c r="J5" s="4"/><c r="K5" s="4"/><c r="L5" s="4"/><c r="M5" s="4"/><c r="N5" s="4"/><c r="O5" s="5" t="s"><v>30</v></c><c r="P5" s="4"/><c r="Q5" s="4"/><c r="R5" s="4"/><c r="S5" s="4"/><c r="T5" s="4"/><c r="U5" s="4"/></row><row r="6"><c r="A6" s="2" t="s"><v>21</v></c><c r="B6" s="2" t="s"><v>22</v></c><c r="C6" s="5" t="s"><v>31</v></c><c r="D6" s="2"><v>5.0</v></c><c r="E6" s="2"><v>2019.0</v></c><c r="F6" s="4"/><c r="G6" s="4"/><c r="H6" s="4"/><c r="I6" s="4"/><c r="J6" s="4"/><c r="K6" s="4"/><c r="L6" s="4"/><c r="M6" s="4"/><c r="N6" s="4"/><c r="O6" s="5" t="s"><v>32</v></c><c r="P6" s="4"/><c r="Q6" s="4"/><c r="R6" s="4"/><c r="S6" s="4"/><c r="T6" s="4"/><c r="U6" s="4"/></row><row r="7"><c r="A7" s="2" t="s"><v>21</v></c><c r="B7" s="2" t="s"><v>22</v></c><c r="C7" s="5" t="s"><v>33</v></c><c r="D7" s="2"><v>6.0</v></c><c r="E7" s="2"><v>2019.0</v></c><c r="F7" s="4"/><c r="G7" s="4"/><c r="H7" s="4"/><c r="I7" s="4"/><c r="J7" s="4"/><c r="K7" s="4"/><c r="L7" s="4"/><c r="M7" s="4"/><c r="N7" s="4"/><c r="O7" s="5" t="s"><v>34</v></c><c r="P7" s="4"/><c r="Q7" s="4"/><c r="R7" s="4"/><c r="S7" s="4"/><c r="T7" s="4"/><c r="U7" s="4"/></row><row r="8"><c r="A8" s="4"/><c r="B8" s="4"/><c r="C8" s="4"/><c r="D8" s="4"/><c r="E8" s="4"/><c r="F8" s="4"/><c r="G8" s="4"/><c r="H8" s="4"/><c r="I8" s="4"/><c r="J8" s="4"/><c r="K8" s="4"/><c r="L8" s="4"/><c r="M8" s="4"/><c r="N8" s="4"/><c r="O8" s="4"/><c r="P8" s="4"/><c r="Q8" s="4"/><c r="R8" s="4"/><c r="S8" s="4"/><c r="T8" s="4"/><c r="U8" s="4"/></row><row r="9"><c r="A9" s="4"/><c r="B9" s="4"/><c r="C9" s="4"/><c r="D9" s="4"/><c r="E9" s="4"/><c r="F9" s="4"/><c r="G9" s="4"/><c r="H9" s="4"/><c r="I9" s="4"/><c r="J9" s="4"/><c r="K9" s="4"/><c r="L9" s="4"/><c r="M9" s="4"/><c r="N9" s="4"/><c r="O9" s="4"/><c r="P9" s="4"/><c r="Q9" s="4"/><c r="R9" s="4"/><c r="S9" s="4"/><c r="T9" s="4"/><c r="U9" s="4"/></row><row r="10"><c r="A10" s="4"/><c r="B10" s="4"/><c r="C10" s="4"/><c r="D10" s="4"/><c r="E10" s="4"/><c r="F10" s="4"/><c r="G10" s="4"/><c r="H10" s="4"/><c r="I10" s="4"/><c r="J10" s="4"/><c r="K10" s="4"/><c r="L10" s="4"/><c r="M10" s="4"/><c r="N10" s="4"/><c r="O10" s="4"/><c r="P10" s="4"/><c r="Q10" s="4"/><c r="R10" s="4"/><c r="S10" s="4"/><c r="T10" s="4"/><c r="U10" s="4"/></row></sheetData><drawing r:id="rId1"/></worksheet>

**Comment by rockandruby** [comment]
@chopraanmol1 any suggestions?

**Comment by chopraanmol1** [comment]
@rockandruby roo doesn't create an XML file by itself rather it extract xml(s) from xlsx file which in this case already have float representation of the integer. Most of Excel client doesn't represent an integer as float while saving the spreadsheet but it seems Google-Spreadsheet is doing that anyway.

It generally still work fines as while reading the xlsx file excel client uses the format of the respective cell to determine how to display the data. In roo's scenario, you can use [formatted_value](https://www.rubydoc.info/gems/roo/2.8.2/Roo/Excelx#formatted_value-instance_method) to do the same, which in the current example will return String wrapped Integer value.

We cannot use format to calculate a cell's value as it may result in data loss and will be considered breaking change.

Implementing solution for this problem (converting all N.0 (float)  => N (Integer))  is quite simple, but the problem is that other user may expect the current result.

I need to think a bit more on this if there is any safe way to typecast without breaking existing scenario Or whether it won't be a breaking change to begin with.

**Comment by rockandruby** [comment]
@rockandruby thanks for reply. Already did monkey patching)https://github.com/roo-rb/roo/issues/498#issuecomment-472330336
